man/man1/lombscargle.1
.TH lombscargle 1 "January 2010" "CEH" \" -*- nroff -*-
.SH NAME 
lombscargle \- Estimate the power spectrum from an unevenly sampled time series.
.br
.SH SYNOPSIS
.B lombscargle [options]
.br

.br
Requires Martin De Kauwe's C Library to run:
.br
#include <memory.h>
.br
#include <file_io.h>
.br
#include <error_handling.h> 
.br 
.SH DESCRIPTION
.br
This code is an implementation of the Lomb-Scargle method which uses least squares 
to fit sine and cosine model functions to an unequally sampled 
time-series, y(t). Use of the least squares method allows the observed time-series 
to be evaluated at measured intervals by weighting on a "per point interval", 
rather than a "per time interval" basis (Press, 1992), i.e. can be used on time
series which contain gaps. 

.p	
Given N observations, y(t), obtained at time t_i (i=1,2,...,N) a model can be 
formed:

	y(t) = A.cos(2.pi.omega) + B.sin(2.pi.omega)

.br	
.p	
where A and B are unknown and omega is the known frequency. Using the method of 
least squares the lomb normalised periodgram is defined as:
	

	P_N(omega) = 0.5.sigma^2 x [ AA^2 / BB + CC^2 / DD ]

.br	
.p
where AA  = SUM (y_tn - ybar).cos(2.pi.omega(tn - tau)),
.br
BB = SUM cos^2(2.pi.omega(tn - tau)),
.br
CC = SUM (y_tn - ybar).sin(2.pi.omega(tn - tau)),
.br
DD = SUM sin^2(2.pi.omega(tn - tau)),
.br
tau = tan(2.omega.tau) 
.br 
    =  SUM (sin(2.omega.tn) / SUM (cos(2.omega.tn),
.br
and tau makes the equation time-invariant.

.p
This code also implements the spectrum corrections presented in Schulz & Nudelsee (2002).
Time-series can show a decrease in spectral amplitude with increasing frequency, 
dubbed "red-noise". This code corrects the periodogram by testing whether the 
variability recorded for a given time-series is consistent with that which 
stochastically arises from a first-order autoregressive process. 

.p
Finally, there is an option to output the input data required to carry out cross-spectral
analysis between two time series. Check manpage for coherency for more details.

.br

.SH HELP OPTIONS
.IP "-u " 9
Print a usage message (options with brief explanation).
.br

.SH GENERAL OPTIONS
.IP "-ver " 9
Print to the screen the program build version, date information.
.br
.IP "-segs \fI\|n\|\fR" 9
Number of WOSA segments with a 50 percent overlap, i.e. -segs 3
.br
.IP "-b " 9
Use binary mode - i.e. code expects as input a text file with the date locations of binary files. You
will also need to supply the dimensions of the binary files, the file extentions etc, but the code 
will do the rest. 
.br
.IP "-dir " 9
Set path to binary files, default is /users/eow/mgdk/DATA/REGRIDDED_LST/gridded_03/
.br
.IP "-fname " 9
 Set file prefix, default is lsta_daily_
.br
.IP "-fext " 9
Set file extension, default is .gra
.br
.IP "-scargle " 9
Generate periodogram using Scargle paper implementation.
.br
.IP "-press " 9
Generate periodogram using the Press et al implementation.
.br
.IP "-my " 9
Average multiyears of input data to get a single timeseries and process spectra of this
.br
.IP "-runs " 9
Check it is appropriate to use an AR1 model to describe the time-series 
Using a non-parametric test according to Bendat and Piersol, 1986, Random
data, 2nd ed, wiley: New York, pg 95.
.br
.IP "-nrows \fI\|n\|\fR" 9
The number of rows in binary image (int).
.br
.IP "-ncols \fI\|n\|\fR" 9
The number of columns in binary image (int).
.br
.IP "-r \fI\|n\|\fR" 9
The row in the binary image you want processed (int). Note code is zero based.
.br
.IP "-c \fI\|n\|\fR" 9
The column in the binary image you want processed (int). Note code is zero based.
.br
.IP "-undef \fI\|n\|\fR" 9
Set undefined value in input file - LST = -999., default is -500. Code expects a 
floating point value.
.br
.IP "-white " 9
Test significance of generated power spectrum against hypothesis of random white noise.
.br
.IP "-si \fI\|n\|\fR" 9
Explicitly set the sampling interval of the output spectra i.e. 0.001. Code expects a float.
.br
.IP "-ai \fI\|n\|\fR" 9
Explicitly set the average sampling interval. Code expects a float.
.br
.IP "-of \fI\|n\|\fR" 9
Oversampling factor, set to 4.0 (typical values >= 4.0). This improves the resolution of the output
spectrum, as the sampling interval is increased. Clearly there is a limit, and one can't produce an
improved resolution where there isn't one already exisiting in the original time series.
.br
.IP "-hi \fI\|n\|\fR" 9
Specify the scaling factor of the average Nyquist frequency, (Hifac x avg.Nyquist freq = Highest 
frequency to be examined) set to 1.0 by default
.br
.br
.IP "-mf \fI\|n\|\fR" 9
Explicitly set the maximum frequency sampled upto, e.g. the nyquist frequency or beyond (though clearly
not recommended).
.br
.IP "-eq " 9
Data is equal distance use standard method for calculating autocorrelation coefficent.
.br

.SH DATA TAPERING OPTIONS
A fourier transform of a time series often results in "spectral leakage", where power at a given frequency
interval contains leakage from another wavelength which is distance X away. This happens because a fourier 
transform ends up with non-zero values at frequenceies other than omega, for example caused by an abrupt end
to the input time series. This can result in it being difficult
to resolve a sinsoidal component. Data windowing is therefore a process to modulate (gradually) the input time 
series.
.IP "-nowin "  9
Don't apply a window to the raw time series.
.br
.IP "-costap "  9
Apply a split cosine bell tapering to the time series.
.br
.IP "-rec "  9
Use a rectangular window on the raw time series.
.br
.IP "-han "  9
Use a Hanning window on the raw time series.
.br

.SH TREND OPTIONS
A prerequisite of harmonic analysis is the need for the data to be stationary (unchanging), if not a 
spectrum becomes very difficult to interpret. Clearly a time series will contain features such as noise, 
so Priestley (1981) argues for the need for "weak stationarity". The default option is to apply a linear 
least squares fit to the data and then subtract the line from the time series.
.IP "-notre " 9
Don't detrend the raw time-series.
.br
.IP "-notre " 9
Detrend using more *robust* fitting a line by minimising absolute deviation method. User beware, the method 
implemented can result in problems - google medfit and see for yourself!
.br

.SH AR1 OPTIONS
To overcome the issue of red-noise bias and obtain an improqved periodogram, a first-order
autoregressive (AR1) process is fitted to the time series. The significance of peaks in
frequency space can therefore be tested against a red-noise background generated by the AR1 process.
.IP "-seed \fI\|n\|\fR" 9
Set the random seed used to generate random numbers for the AR1 process. You must specify a negative int.
.br
.IP "-nsim \fI\|n\|\fR" 9
The number of AR1 simulations to generate, default is 200.
.br
.IP "-fconf99 " 9
Output significance at the 99% level.
.br
.IP "-fconf95 " 9
Output significance at the 95% level.
.br
.IP "-fconf90 " 9
Output significance at the 90% level.
.br
.IP "-fconf80 " 9
Output significance at the 80% level.
.br

.SH OUTPUT OPTIONS
.IP "-dumpin " 9
Dump input file out to stderr. Ok sounds pointless, but it is useful if you are processing a binary file.
.br	
.IP "-year \fI\|n\|\fR" 9
Dump input file out to stderr, but determine year of interest, i.e. if looking at 2007, stops output running
from 142-284, it would go from 1-142 instead.
.br	
.IP "-printo " 9
Print corrected (red noise) and uncorrected periodgrams.
.br	
.IP "-printr " 9
Print periodgrams, in relative power on the yaxis and period on xaxis.
.br	
.IP "-fp " 9
Find the statistically significant peaks and output the bandwidth resolution, plot in gnuplot u  1:5:6 w xerr.
.br	
.IP "-fp " 9
Print periodgrams in a decibel scale.
.br	

.SH INPUT FILE FORMAT
The input file format is either an a ascii file provided on the standard input,
or an ascii file with the locations of various binary input files. If the ascii 
format is supplied, then the code expects two columns) - i.e: 
		
		time 1    value 1
.br
		time 2    value 2
.br
		  .          .
.br
		  .          .          
.br
		time n     value n


Alternatively on the stdin a ascii text file with the date stamp of the relevant 
binary files requiring processing. If using binary mode (-b) the number of rows, 
number of cols, row and col of interest need to also be defined on the command line.
Similarly the path to the file (-dir), the file type (-fname) and the file extention
(-fext), though defaults are listed.
 
Note the code only works on a single pixel!

.p	
for example:
 -segs 3
.br
lombscargle -b -dir "/users/eow/mgdk/DATA/REGRIDDED_LST/gridded_03/" \
			-fname "lsta_daily_" -fext ".gra" -segs 3 -r 100 -c 100 -nrows 348 \
			-ncols 728 < input_files > output_file
.br	
where the contents of input_files might look like:

.br	
20060522
20060523
20060524

.SH EXAMPLE
run the shell:

.br
example_doMe.bash 

.SH OUTPUT FILE FORMAT
The default form of the output file is 4 colums of data, one row per frequency interval.
Column headings are in order:

.br 

\fI\|frequency\|\fR (cycles per time interval);
.br
\fI\|power\|\fR (unitless);
.br
\fI\|significance level\|\fR (unitless);
.br
\fI\|scaler\|\fR (unitless, to turn the power into relative power (i.e. divide column 2 by column 4));
.br

The output is supressed if the user requests the "-dumpin" flag.
In such a scenario the input file is printed instead.
.br

.SH REFERENCES
.br
Chatfield, C. (2004) The analysis of time series - an introduction. Chapman and Hall,
Boca Raton. Pg 159-167.
.br   

Glynn, E. F., Chen, J. and Mushegian, A. R. (2005) Detecting periodic patterns in unevenly
spaced gene expression time series using Lomb-Scargle periodograms. Bioinformatics, 22, 310-316.
.br

Lomb, N. R. (1976) Least-squares frequency analysis of unequally spaced data. Astrophysics
and Space Science, 39, 447-462.
.br

Press, W. H., Teukolsky, S. A., Vetterling, W. T. and Flannery, B. P. (1992) Numerical 
Recipes in C: The art of scientific computing. Cambridge University Press, Cambridge.
.br

Priestley, M. B. (1981) Spectral analysis and time series. Academic Press, London.
.br

Scargle, J. D. (1982) Statistical aspects of spectral analysis of unevenly spaced data. 
Astrophysics Physics Journal, 263, 835-853.
.br

Schulz & Mudelsee (2002) REDFIT: estimating red-noise spectra directly from unevenly spaced 
paleoclimatic time series. Computers & Geosciences, 28, 421-426.
.br    

.SH BUGS
None as we speak, yeah right!
.SH SEE ALSO
coherency(1)
.br
.SH AUTHOR/COPYRIGHT
.br
Martin De Kauwe
.br
Centre of Ecology and Hydrology
.br
Crowmarsh Gifford
.br
Wallingford
.br
Gower Street
.br
Oxfordshire, OX10 8BB 
.br
mgdk@ceh.ac.uk
.br
mdekauwe@gmail.com
.br

.br
Correction for rednoise is based on the fortran code REDFIT:
.br
M. Schulz and M. Mudelsee
.br

.br
Original version of period code:
.br
Press et al 1992
.br


lombscargle.c
/*========================================================================================
    Lomb-Scargle algorithm    - Estimate the power spectrum from an unevenly sampled time 
                              series, with red noise bias correction or potential to test
                              against white noise hypothesis.
  ========================================================================================
    
    ++ READ THE MANPAGE, CHANGES.TEXT
    ++ Code depends on being able to link against my C library functions...
        
     Martin De Kauwe: 13th January, 2010                                                         
=========================================================================================*/

#include "lombscargle.h"

int main(int argc, char **argv)
{
    /* setup up space to pass stuff around in */
    control *c;
    if ((c = malloc(sizeof(control))) == NULL) {
        check_errors("Control structure: Not allocated enough memory", __LINE__);
     }
    
    timeseries *ts;
    if ((ts = malloc(sizeof(timeseries))) == NULL) {
        check_errors("Timeseries structure: Not allocated enough memory", __LINE__);
     }
    
    meta *m;
    if ((m = malloc(sizeof(meta))) == NULL) {
        check_errors("Meta structure: Not allocated enough memory", __LINE__);
     }
    
    output_options *o;
    if ((o = malloc(sizeof(output_options))) == NULL) {
        check_errors("Output_options structure: Not allocated enough memory", __LINE__);
     }
    
    /* setup initial junk */
    setup_initial_conditions(c, ts, o);
    
    /* send sweep along the command line */
    clparse(argc, argv, c, ts, o);
    
    /* check the input file is on the stdin? */
    check_stdin(__LINE__);
    
    /* read data in from stdin */
    read_input_file(argv, c, ts, o);
    
    /* setup space for working arrays and other misc stuff */
    allocate_working_environment(c, ts);
    
    /* calculate lomb periodgram for raw time-series */
    spectrum_wrapper(c, ts, ts->x, ts->y, ts->frq, ts->gxx);
    
    /* Calculate red noise bias and correct periodgram 
        - need to generate a seed ONCE */
    srand (time(NULL));
    rednoise_bias_wrapper(c, ts, o);
    
    /* dump out what you need */
    print_outputs(c, ts, o);
    
    /* tidy up */
    free_array_double(ts->gxx);
    free_array_double(ts->frq);
    free_array_double(ts->red);
    free_array_double(ts->grr);
    free_array_double(ts->grrsum);
    free_array_double(ts->gredth);
    free_array_double(ts->corr);
    free_array_double(ts->gxxc);
    free_array_double(ts->grravg);
    free_array_double(ts->x);
    free_array_double(ts->y);  
    free(c);
    free(ts);
    free(m);
    free(o);
    
    return (0);
}

void spectrum_wrapper(control *c, timeseries *ts, double *xdata, double *ydata, double *frequency, double *power)
{
    /* 
        wrapper func - to calculate periodgram for given time-series 
        also has the options to remove trends from the data and apply
        a window.
    */
    
    int     i, j, istart;
    double  intercept = 0., abdev = 0., slope = 0.;
    double *ftrx = NULL, *ftix = NULL, *xwk = NULL, *ywk = NULL, scal;
    
    ftrx = allocate_memory_double(ts->lfreq+1, __LINE__);
    ftix = allocate_memory_double(ts->lfreq+1, __LINE__);
    xwk  = allocate_memory_double(ts->nseg+1,  __LINE__); /* periodgram power */
    ywk  = allocate_memory_double(ts->nseg+1,  __LINE__); /* corresponding frequencies */
    
    /* scale autospectrum and setup frequency axis */
    scal = 2.0 / ((double)ts->n50 * (double)ts->nseg * ts->df * ts->ofac);
    ts->factor = scal;    /* store for coherency code */
             
    for (i = 1; i <= ts->n50; i++) {
        
        /* copy data of i'th segment into workspace */
        istart = (int)((i - 1) * (double)ts->nseg / 2.);
        for (j = 1; j <= ts->nseg; j++) {
            xwk[j] = xdata[istart + j];
              ywk[j] = ydata[istart + j];
            
        }
    
        /* detrend the data */
        if (c->DETREND == TRUE) {
            rmtrend(c, ts, ywk, xwk);
        } else if (c->ROBUSTDETREND == TRUE) {
            /* 
                fitting method that is more sensitive to outliers in 
                the time-series, I guess the way to go? 
            */
            medfit(ts->nseg, ywk, xwk, &intercept, &slope, &abdev);
    
            /* Subtract the fitted line from the time-series to obtain a linear detrend */
            for (j = 1; j <= ts->nseg; j++) {
                ywk[j] -= (slope * xwk[j]) + intercept;
            }
        }       
        
        /* apply window to data */
        if (c->WINDOW_TYPE != NO_WINDOW) {
            if (c->WINDOW_TYPE == COSINE_TAPER) {
                taper_timeseries(c, ts, ywk);    /* leads to some bizarre results if used in coherency */
            } else {
                window(c, ts, ywk, xwk);
            }
        }
    
        /* 
            calculate periodgram 
                - there are two options the original implementation in Scargle paper
                  or the Press et al method. They give identical results.
        */
        if (c->PERIODOGRAM_TYPE == SCARGLE) { 
            ft_uneven_data(xwk, ywk, ftrx, ftix, ts->nfreq, ts->nseg, ts->lfreq, ts->wz);
            
            /* sum raw spectra */
            for (j = 1; j <= ts->nout; j++) {
                power[j] += ftrx[j] * ftrx[j] + ftix[j] * ftix[j];
                
            }
        } else if (c->PERIODOGRAM_TYPE == PRESS) {
            period(xwk, ywk, ftrx, ftix, ts->nseg, ts->df, ts->nout);
            
            /* sum raw spectra */
            for (j = 1; j <= ts->nout; j++) {
                power[j] += ftrx[j] * ftrx[j] + ftix[j] * ftix[j];
            }
        }
        for (j = 1; j <= ts->nseg; j++) {
            xwk[j] = 0.0;
              ywk[j] = 0.0;
        }    
    }    
    
    /* rescale periodogram */
    if (c->PERIODOGRAM_TYPE == SCARGLE) { 
        for (i = 1; i <= ts->nout; i++) {
            power[i] *= scal;
            frequency[i] = (i - 1) * ts->df;
        }
    } else if (c->PERIODOGRAM_TYPE == PRESS) {
        for (i = 1; i <= ts->nout; i++) {
            power[i] *= scal;
            /* press algorithm starts at df, rather than 0.0 */
            frequency[i] = i * ts->df;
        }
    }
    
    /* tidy up */
    free_array_double(xwk);   
    free_array_double(ywk);
    free_array_double(ftrx);
    free_array_double(ftix);
    
    return;
}

void rednoise_bias_wrapper(control *c, timeseries *ts, output_options *o)
{
    /* 
        As mentioned already, spectra of time-series shows a decrease in
        spectral amplitude with increasing frequency - red-noise. By fitting
        a first-order autoregressive (AR1) process to the irregularly spaced time-series
        and transforming this to the frequency domain, the spectra can be 
        tested against the hypothesis that the time-series originates from an AR1
        process.
        
        See Schulz and Mudelsee (2002). 
    */
    int    i, k, rprv, rnow; 
    double sum = 0.0, fac90, fac, rho, rhosq, fnyq;

    /* estimate variance for raw periodgram - needed to correct data */
    for (i = 1; i <= ts->nout; i++) {
          sum += ts->gxx[i];
    }
    ts->var = ts->df * sum;
    
    if (c->EQUAL_DIST_DATA == TRUE) {
        ts->equalrho = calculate_persistence(ts);
    } else {
        /* Estimate tau - unless prescribed? */
        if (ts->rho < 0.0) {
            /* Estimate a value for tau */
            gettau(c, ts);
            /* can't have a negative tau */
            if (ts->tau < 0.0) {
                ts->tau = 0.0;
                fprintf(stderr, "Negative tau returned, tau forced to zero: %f\n", ts->tau);
            }    
        } else {
            ts->tau = -(ts->avgdt) / log(ts->rho);
        }
    }     
    
    /* 
        need to feed random numbers with a different seed - one off runs can simply use the clock.
        However, when running on multiple machines, there is potentially a scenerio where the same
        seed will be used, hence there is the option to feed the program with a seed on the cmd
        line. Ideally some bash func that generates a load of NEGATIVE ints.
    */
    if (c->CMD_LINE_SEED == FALSE) {
        ts->idum = -fabs(rand());
    } else {
        if (ts->seed > 0) {
            check_errors("Error: You must seed me with a negative int", __LINE__);
        } else {
            ts->idum = ts->seed;
        }
    }
    
    /* if debugging...*/
    /*ts->idum = -123456789;*/
    
    /* 
        As the lomb-scargle spectrum cannot be directly corrected (dependent on 
        the sampling intervals), a monte-Carlo ensemble, nsim AR1 time-series are generated.
        The difference between the average ensemble spectrum from the theoretical
        specturm is used as the correction.
    */
    
    /* gneratre AR(1) spectra */
    for (i = 1; i <= ts->nsim; i++) {
        
        /* set up AR1 junk */
        make_AR1(c, ts);
        
        /* calculate lomb periodgram for AR1 spectra */
        spectrum_wrapper(c, ts, ts->x, ts->red, ts->frq, ts->grr);
        
        /* scale and sum red-noise spectra */
        sum = 0.0;
        for (k = 1; k <= ts->nout; k++) {
            sum += ts->grr[k];
        }
        
        ts->varx = ts->df * sum;
        fac      = ts->var / ts->varx;
        for (k = 1; k <= ts->nout; k++) {
            ts->grrsum[k] += ts->grr[k] * fac;    
        }
        
        /* Need to zero grr array */
        for (k = 1; k <= ts->nout; k++) {
            ts->grr[k] = 0.0;
        }
            
    }
    
    /* average red-noise spectrum */
    sum = 0.0;
    for (k = 1; k <= ts->nout; k++) {
        ts->grravg[k] = ts->grrsum[k] / ((double)ts->nsim);
        sum += ts->grravg[k]; 
    }
    
    ts->varx = ts->df * sum;
    fac = ts->var / ts->varx;
    
    for (k = 1; k <= ts->nout; k++) {
        ts->grravg[k] *= fac;
    }
    
    /* determine lag-1 autocorrelation coeff, which decays exponetially as a func of time */
    if (c->EQUAL_DIST_DATA == TRUE) {
        rho   = ts->equalrho; 
        rhosq = rho * rho;
    } else {
        rho   = exp((-1.0 * ts->avgdt) / ts->tau);
        rhosq = rho * rho;
    }
    
    
    /* 
        If we want to test the spectrum against the background white noise
        spectrum instead then according to Mann and Lees (1996) Eq. 1, then setting
        rho to zero yields a white-noise process, I guess technically we also need to
        set rho in the AR1 stuff, but this seems fine as gredth is what we use to form
        the significance level 
    */
    if (c->SIGNIF_TEST == WHITE) {
        rho = 0.0;
    }
    
    /* 
        set theoretical spectrum (e.g., Mann and Lees, 1996, Eq. 4) 
         make area equal to that of the input time series
    */
    fnyq = ts->frq[ts->nout];    
           
      sum  = 0.0;
    for (k = 1; k <= ts->nout; k++) {
        /* power spectrum of an AR(1) process is given by ... */
        ts->gredth[k] = (1.0 - rhosq) / (1.0 - 2.0 * rho * cos(M_PI * ts->frq[k] / fnyq) + rhosq);
        sum += ts->gredth[k];         
      }
    
    ts->varx = ts->df * sum;
      fac = ts->var / ts->varx;
    
    /* 
        work out degrees of freedom and signif. The degrees of freedom affect the size of the 
        confidence levels. Greater dof = a smoother spectrum, the lower the variance of the estimates 
        and the narrower the confidence intervals and hence levels (Weedon 2003). 
    */
    getdof(c, ts);
    fac90 = getchi2(ts->dof, ts->alpha) / ts->dof;
    
    /* determine correction factor */
    for (k = 1; k <= ts->nout; k++) {
        ts->gredth[k] *= fac;
        ts->corr[k] = ts->grravg[k] / ts->gredth[k];
        ts->gxxc[k] = ts->gxx[k] / ts->corr[k];
        ts->gredth[k] *= fac90;    
        
        /* store up area under power spectrum to normalise by */
        ts->rel_pow_rescale += ts->gxxc[k];    
    }

    if (c->DO_RUNS_TEST == TRUE) {
        /* 
            Check it is appropriate to use an AR1 model to describe the time-series 
            Using a non-parametric test according to Bendat and Piersol, 1986, Random
            data, 2nd ed, wiley: New York, pg 95.
        */
        if ((c->WINDOW_TYPE != RECTANGULAR) || (ts->ofac >= 1.5) || (ts->n50 != 1)) {
            check_errors("You can only apply the runs test using a rectangular window, ofac = 1.0, and segs = 1", __LINE__);
        }
    
        ts->runs_count = 1;
        rprv = SIGN(1.0, ts->gxxc[1] - ts->gredth[1]);
 
        for (k = 1; k <= ts->nout; k++) {
            rnow = SIGN(1.0, ts->gxxc[k] - ts->gredth[k]);
            if (rnow != rprv) {
                ts->runs_count++;
            }
            rprv = rnow;
        }
    
        /* note I have added 0.5 at the end to make sure rounding when converting to int is right */
        ts->rcrithlo = (int)((pow((-0.79557086 + 1.0088719 * sqrt((double)(ts->nout/2))), 2.0)) + 0.5);
        ts->rcrithi  = (int)((pow(( 0.75751462 + 0.9955133 * sqrt((double)(ts->nout/2))), 2.0)) + 0.5);
        
        print_outputs(c, ts, o);
    }

    return;
}

void setup_initial_conditions(control *c, timeseries *ts, output_options *o)
{
    strcpy(c->versionstamp, "Lombscargle V2.0, Build Date: " __DATE__ " " __TIME__ ", Martin De Kauwe\n");
    c->DO_RUNS_TEST       =  FALSE;
    ts->factor            =  0.0;
    ts->undef_val         = -500.0;        /* undef values in lst are -999., so 500 seems a nice number! */
    c->AVERAGE_MULTI_YEAR =  FALSE;
    c->EQUAL_DIST_DATA    =  FALSE;
    c->AUTO_SET_SAMP_INT  =  FALSE;
    c->CMDAVGDT           =  FALSE;
    c->CMDSI              =  FALSE;
    c->WHAT_IS_NYQ        =  FALSE;
    c->CMD_LINE_SEED      =  FALSE;
    c->WINDOW_TYPE        =  HANNING;
    c->DETREND            =  TRUE;
    c->ROBUSTDETREND      =  FALSE;
    c->ENOUGH_DATA        =  FALSE;
    c->INPUT_FILE_TYPE    =  ASCII;
    c->SIGNIF_TEST        =  RED;
    c->PERIODOGRAM_TYPE   =  SCARGLE;    /* option to use scargle implementation rather than Press et al method */
    ts->num_param         =  0;
    ts->ofac              =  4.0;  /* oversampling frequency */
    ts->hifac             =  1.0;     /* Specify the scaling factor of the average Nyquist frequency */
    ts->min_interval      =  0.0;
    ts->rel_pow_rescale   =  0.0;   /* used to rescale power spectrum to be relative power */
    ts->min_err           =  0;
    ts->n50               =  1;
    c->ifp                =  stdin;
    ts->nsim              =  200;
    ts->alpha             =  CONF90;
    c->row                = -99;   /* set to be negative so we can check it is set on cmd line if code used in binary mode */ 
    c->start_row          = -99;   /* set to be negative so we can check it is set on cmd line if code used in binary mode */ 
    c->end_row            = -99;   /* set to be negative so we can check it is set on cmd line if code used in binary mode */ 
    c->start_col          = -99;   /* set to be negative so we can check it is set on cmd line if code used in binary mode */ 
    c->end_col            = -99;   /* set to be negative so we can check it is set on cmd line if code used in binary mode */ 
    ts->sampling_int      = -99;   /* set to be negative so we can check it is set on cmd line */ 
    c->num_frames         =  0;
    c->USE_SET_MAX_FREQ   =  FALSE;
    ts->rho               = -99.;
    c->year_of_interest   = -99;
    
    /* Set default input directory */
    strcpy(c->dir_fn,   "/users/eow/mgdk/DATA/REGRIDDED_LST/gridded_03/");
    strcpy(c->file_fn,  "lsta_daily_");
    strcpy(c->file_ext, ".gra");
    
    /* output options */
    o->OUTPUT_TYPE = PRINT_FREQ_SCALE;
    
    return;
}




void make_AR1(control *c, timeseries *ts)
{
    /* 
        set up AR1 time series. An AR spectrum models the data using a linear combination of the previous and 
        subsequent time steps (see Mann and Lees, 1996): 
        
                red[i] = rho[i] * y[i-1] + eps[i]     

           where
                                t[i] - t[i-1]
                rho[i] =  exp(--------------)
                                   tau

    
        and 
        
                 eps ~ NV(0, vareps). To ensure Var[red] = 1, we set

                                    2 * (t[i] - t[i-1])
                vareps = 1 -  exp(- -------------------).
                                      tau

        Stationarity of the generated AR(1) time series is assured by dropping
        the first N generated points.
    */
    
    int    k;
    double dt, sigma;
    
    ts->red[1] = gasdev(&ts->idum);
    
    for (k = 2; k <= ts->num_param; k++) {
        dt = ts->x[k] - ts->x[k - 1];
        sigma = sqrt(1.0 - exp(-2.0 * dt / ts->tau));
        if (c->EQUAL_DIST_DATA == TRUE) {
            ts->red[k] = exp(-dt / ts->equalrho) * ts->red[k - 1] + sigma * gasdev(&ts->idum);
        } else { 
            ts->red[k] = exp(-dt / ts->tau) * ts->red[k - 1] + sigma * gasdev(&ts->idum);
        }
    }
    
    return;
}

void allocate_working_environment(control *c, timeseries *ts)
{
    long   i;
    double tp, sumdt = 0.;
    
     ts->nseg = (int)(2. * ts->num_param / (ts->n50 + 1)); /* pts per segment */ 
    
    /* average sampling interval of entire time series */
    for (i = 2; i <= ts->num_param; i++) {
        sumdt += ts->x[i] - ts->x[i-1];
        
    }
    
    /* set sampling int on cmd line */
    if (c->CMDAVGDT == FALSE) {
        ts->avgdt = sumdt / ((double)(ts->num_param - 1.0));
    } 
    
    tp = ts->avgdt * ((double)ts->nseg); /* average period of a segment  */
    
    if (c->AUTO_SET_SAMP_INT == TRUE) {
        if ((ts->avgdt > 0.999999) && (ts->avgdt <= 1.00001)) {
            c->EQUAL_DIST_DATA = TRUE;
        }
    }
    
    if (c->CMDSI == FALSE) { 
        ts->df = 1.0 / (ts->ofac * tp); /* freq. spacing */
    }
    
    ts->wz = 2.0 * M_PI * ts->df; /* omega = 2*pi*f    */
      
    /* 
        the nyquist freq is not well defined for irregularly spaced data
        so use so use average samp interval 
    */
    if (c->USE_SET_MAX_FREQ == FALSE) {
        ts->fnyq = ts->hifac * 1.0 / (2.0 * ts->avgdt); /* average Nyquist freq. */
    } else {
        ts->hifac = ts->max_freq / ( 1.0 / (2.0 * ts->avgdt) );
        ts->fnyq  = ts->hifac * 1.0 / (2.0 * ts->avgdt); /* NYQ freq explictly set */
    }
      
    if (c->WHAT_IS_NYQ == TRUE) {
        printf("%f %f\n", ts->fnyq, ts->df);
        exit(0);
    }
    
    /* seem that the adding of one is to overcome some fortran rounding issue to an int, so isn't required! */
    /*ts->nfreq = (int)(ts->fnyq / ts->df + 1.); */
    ts->nfreq = (int)(ts->fnyq / ts->df); /* number of freq for transform f(1) = f0; f(nfreq) = fNyq */
    
    /* make sure that number of frequencies is even - maybe put this in???? */
    /*if (ts->nfreq % 2 != 0) {
        ts->nfreq--;
    }*/
    ts->lfreq = (int)(ts->nfreq * 2);
      
    /*ts->nout = 0.5 *ts->ofac *ts->hifac * np;*/
    ts->nout  = ts->nfreq;
    
    /* sort out some space */
    ts->gxx    = allocate_memory_double(ts->nout+1,      __LINE__); /* periodgram power */
    ts->frq    = allocate_memory_double(ts->nout+1,      __LINE__); /* corresponding frequencies */
    ts->red    = allocate_memory_double(ts->num_param+1, __LINE__);
    ts->grr    = allocate_memory_double(ts->nout+1,      __LINE__);
    ts->grrsum = allocate_memory_double(ts->nout+1,      __LINE__);
    ts->gredth = allocate_memory_double(ts->nout+1,      __LINE__);
    ts->corr   = allocate_memory_double(ts->nout+1,      __LINE__);
    ts->gxxc   = allocate_memory_double(ts->nout+1,      __LINE__);
    ts->grravg = allocate_memory_double(ts->nout+1,      __LINE__);

    return;
}

